pkg/volume/git_repo/git_repo.go

```diff
@@ -138,8 +140,15 @@ func (b *gitRepoVolumeBuilder) SetUpAt(dir string) error {
 		return err
 	}
 
-	if output, err := b.execCommand("git", []string{"clone", b.source}, dir); err != nil {
-		return fmt.Errorf("failed to exec 'git clone %s': %s: %v", b.source, output, err)
+	// If provided target directory
+	if "" != b.target {
+		if output, err := b.execCommand("git", []string{"clone", b.source, b.target}, dir); err != nil {
+			return fmt.Errorf("failed to exec 'git clone %s %s': %s: %v", b.source, b.target, output, err)
+		}
+	} else {
+		if output, err := b.execCommand("git", []string{"clone", b.source}, dir); err != nil {
+			return fmt.Errorf("failed to exec 'git clone %s': %s: %v", b.source, output, err)
+		}
 	}
```

Comment:
As per #15421

How urgent is this one small feature? I'm not deeply against adding it if
it solves real problems, but I'd rather focus energy on the git-sync
sidecar. Additionally, we need clarity on whether this is just "checkout
and rename" or whether "." is expected to be supported. Doing a checkout to
a subdir has one significant advantage - it can be updated atomically. This
means you could periodically sync to head of a branch, which seems like a
powerful feature to me.

If we support ".", that does not work.

On Wed, Oct 14, 2015 at 12:54 AM, Harry Zhang <notifications@github.com>
wrote:

> In pkg/volume/git_repo/git_repo.go
> <https://github.com/kubernetes/kubernetes/pull/15533#discussion_r41964131>
> :
>
> > @@ -138,8 +140,15 @@ func (b *gitRepoVolumeBuilder) SetUpAt(dir string) error {
> >  		return err
> >  	}
> >
> > -	if output, err := b.execCommand("git", []string{"clone", b.source}, dir); err != nil {
> > -		return fmt.Errorf("failed to exec 'git clone %s': %s: %v", b.source, output, err)
> > +	// If provided target directory
> > +	if "" != b.target {
> > +		if output, err := b.execCommand("git", []string{"clone", b.source, b.target}, dir); err != nil {
> > +			return fmt.Errorf("failed to exec 'git clone %s %s': %s: %v", b.source, b.target, output, err)
> > +		}
> > +	} else {
> > +		if output, err := b.execCommand("git", []string{"clone", b.source}, dir); err != nil {
> > +			return fmt.Errorf("failed to exec 'git clone %s': %s: %v", b.source, output, err)
> > +		}
> >  	}
>
> This will fail when clone directory is '.'
>
> Actually, see my comments here: #15421 (comment)
> <https://github.com/kubernetes/kubernetes/issues/15421#issuecomment-147963960>
>
> I'd like to update the code if we reached a agreement.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/15533/files#r41964131>.
>